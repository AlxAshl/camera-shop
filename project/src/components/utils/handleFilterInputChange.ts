import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';

export const handleFilterInputChange = (filterName: string, filter: string[], action: ActionCreatorWithNonInferrablePayload) => {
  const mutableFilter = Object.assign([], filter);
  if(filter.includes(filterName)) {
    const updatedFilter = mutableFilter.filter((entry) => entry !== filterName);
    store.dispatch(action(updatedFilter));
  }
  else {
    mutableFilter.push(filterName);
    store.dispatch(action(mutableFilter));
  }
};
