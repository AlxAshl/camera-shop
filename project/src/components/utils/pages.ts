import { PAGE_LIMIT } from '../../const';

function getPagesCount(cameraCount: number): number {
  return Math.ceil(cameraCount / PAGE_LIMIT);
}

export default getPagesCount;
