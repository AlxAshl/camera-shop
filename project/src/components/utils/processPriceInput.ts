import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';
import { ProductType } from '../../types/product';

export const setMinInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMinPriceInput;
    const newMin = productsArray.find((product) => product.price <= Number(input));
    newMin?.price === undefined
      ? newMinPriceInput = productsArray[productsArray.length - 1].price
      : newMinPriceInput = newMin?.price;
    store.dispatch(action([String(newMinPriceInput)]));
  }
};

export const setMaxInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMaxPriceInput;
    const newMax = productsArray.find((product) => product.price >= Number(input));
    newMax?.price === undefined
      ? newMaxPriceInput = productsArray[productsArray.length - 1].price
      : newMaxPriceInput = newMax?.price;
    store.dispatch(action([String(newMaxPriceInput)]));
  }
};
