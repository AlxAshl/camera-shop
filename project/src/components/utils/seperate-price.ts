export function seperatePrice(priceValue: number) {
  let returnedPrice = String(priceValue);
  if(returnedPrice.length > 6) {
    returnedPrice = `${returnedPrice.slice(0, 1)} ${returnedPrice.slice(1, 4)} ${returnedPrice.slice(-3)}`;
  }
  if (returnedPrice.length > 3) {
    returnedPrice = `${returnedPrice.slice(-0, -3)} ${returnedPrice.slice(-3)}`;
  }
  return returnedPrice;
}
