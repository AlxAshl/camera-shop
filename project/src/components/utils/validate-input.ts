export function validateInput(target: (EventTarget & HTMLInputElement) | (EventTarget & HTMLTextAreaElement)) {

  if (target.parentElement?.parentElement && target.name === 'rating' && (Number(target.value) > 0)) {
    if (!target.parentElement.parentElement.parentElement?.className.includes('is-valid')) {
      target.parentElement.parentElement.parentElement?.classList.add('is-valid');
      target.parentElement.parentElement.parentElement?.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name !== 'review' && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length > 0) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string'){
    if(target.value.length === 0) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length >= 5) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name === 'review' && target.value.length < 5) {
    if (!target.parentElement.parentElement.className.includes('is-invalid')) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

}
