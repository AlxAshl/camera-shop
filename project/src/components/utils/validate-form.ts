import { reviewDataType } from '../../types/review';
import { validateInput } from './validate-input';


export function validateForm(data: reviewDataType) {
  let validationStatus = true;
  Object.entries(data).forEach(([key, value]) =>{
    if (key !== 'review') {
      if(value === '') {
        const failedInput = document.querySelector(`input[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
        validationStatus = false;
      }
    }
    if(key === 'review') {
      if(String(value).length < 5) {
        const failedInput = document.querySelector(`textarea[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
      }
    }
    if(key === 'rating') {
      if(value < 1) {
        const failedInput = document.querySelector(`input[name=${key}]`);
        if(failedInput && failedInput.parentElement?.parentElement){
          failedInput.parentElement.parentElement.parentElement?.classList.add('is-invalid');
        }
        validationStatus = false;
      }
    }
  });
  return validationStatus;
}

