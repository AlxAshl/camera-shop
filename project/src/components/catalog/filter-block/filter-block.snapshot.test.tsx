import { renderWithProviders } from '../../../test/utils/render-with-redux';
import FilterBlock from './filter-block';


test('FilterBlock renders correctly', () => {
  const { asFragment } = renderWithProviders(<FilterBlock/>);
  expect(asFragment()).toMatchSnapshot();
});

