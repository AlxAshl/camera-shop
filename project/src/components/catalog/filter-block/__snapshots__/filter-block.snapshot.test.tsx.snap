// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBlock renders correctly 1`] = `
<DocumentFragment>
  <div
    class="catalog__aside"
  >
    <div
      class="catalog-filter"
    >
      <form
        action="#"
      >
        <h2
          class="visually-hidden"
        >
          Фильтр
        </h2>
        <fieldset
          class="catalog-filter__block"
        >
          <legend
            class="title title--h5"
          >
            Цена, ₽
          </legend>
          <div
            class="catalog-filter__price-range"
          >
            <div
              class="custom-input"
            >
              <label>
                <input
                  data-testid="minPrice-test"
                  id="minPrice"
                  name="price"
                  placeholder="от"
                  type="number"
                  value=""
                />
              </label>
            </div>
            <div
              class="custom-input"
            >
              <label>
                <input
                  data-testid="maxPrice-test"
                  id="maxPrice"
                  name="priceUp"
                  placeholder="до"
                  type="number"
                  value=""
                />
              </label>
            </div>
          </div>
        </fieldset>
        <fieldset
          class="catalog-filter__block"
        >
          <legend
            class="title title--h5"
          >
            Категория
          </legend>
          <div
            class="custom-checkbox catalog-filter__item"
          >
            <label>
              <input
                data-testid="photocamera-test"
                name="photocamera"
                type="checkbox"
              />
              <span
                class="custom-checkbox__icon"
              />
              <span
                class="custom-checkbox__label"
              >
                Фотокамера
              </span>
            </label>
          </div>
          <div
            class="custom-checkbox catalog-filter__item"
          >
            <label>
              <input
                data-testid="videocamera-test"
                name="videocamera"
                type="checkbox"
              />
              <span
                class="custom-checkbox__icon"
              />
              <span
                class="custom-checkbox__label"
              >
                Видеокамера
              </span>
            </label>
          </div>
        </fieldset>
        <fieldset
          class="catalog-filter__block"
        >
          <legend
            class="title title--h5"
          >
            Тип камеры
          </legend>
          <div
            class="custom-checkbox catalog-filter__item"
          >
            <label>
              <input
                data-testid="digital-test"
                name="digital"
                type="checkbox"
              />
              <span
                class="custom-checkbox__icon"
              />
              <span
                class="custom-checkbox__label"
              >
                Цифровая
              </span>
            </label>
          </div>
          <div
            class="custom-checkbox catalog-filter__item"
          >
            <label>
              <input
                data-testid="film-test"
                name="film"
                type="checkbox"
              />
              <span
                class="custom-checkbox__icon"
              />
              <span
                class="custom-checkbox__label"
              >
                Плёночная
              </span>
            </label>
          </div>
          <div
            class="custom-checkbox catalog-filter__item"
          >
            <label>
              <input
                data-testid="snapshot-test"
                name="snapshot"
                type="checkbox"
              />
              <span
                class="custom-checkbox__icon"
              />
              <span
                class="custom-checkbox__label"
              >
                Моментальная
              </span>
            </label>
          </div>
          <div
            class="custom-checkbox catalog-filter__item"
          >
            <label>
              <input
                data-testid="collection-test"
                name="collection"
                type="checkbox"
              />
              <span
                class="custom-checkbox__icon"
              />
              <span
                class="custom-checkbox__label"
              >
                Коллекционная
              </span>
            </label>
          </div>
        </fieldset>
        <fieldset
          class="catalog-filter__block"
        >
          <legend
            class="title title--h5"
          >
            Уровень
          </legend>
          <div
            class="custom-checkbox catalog-filter__item"
          >
            <label>
              <input
                data-testid="level-zero-test"
                name="zero"
                type="checkbox"
              />
              <span
                class="custom-checkbox__icon"
              />
              <span
                class="custom-checkbox__label"
              >
                Нулевой
              </span>
            </label>
          </div>
          <div
            class="custom-checkbox catalog-filter__item"
          >
            <label>
              <input
                data-testid="level-amateur-test"
                name="non-professional"
                type="checkbox"
              />
              <span
                class="custom-checkbox__icon"
              />
              <span
                class="custom-checkbox__label"
              >
                Любительский
              </span>
            </label>
          </div>
          <div
            class="custom-checkbox catalog-filter__item"
          >
            <label>
              <input
                data-testid="level-pro-test"
                name="professional"
                type="checkbox"
              />
              <span
                class="custom-checkbox__icon"
              />
              <span
                class="custom-checkbox__label"
              >
                Профессиональный
              </span>
            </label>
          </div>
        </fieldset>
        <button
          class="btn catalog-filter__reset-btn"
          data-testid="reset-button-test"
          type="reset"
        >
          Сбросить фильтры
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
